src/uwtools/drivers/mpas.py

```diff
@@ -0,0 +1,167 @@
+"""
+A driver for the MPAS component.
+"""
+
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import Optional
+
+from iotaa import asset, dryrun, task, tasks
+
+from uwtools.api.template import render
+from uwtools.config.formats.nml import NMLConfig
+from uwtools.drivers.driver import Driver
+from uwtools.strings import STR
+from uwtools.utils.tasks import file, filecopy, symlink
+
+
+class MPAS(Driver):
+    """
+    A driver for MPAS.
+    """
+
+    def __init__(
+        self,
+        cycle: datetime,
+        config: Optional[Path] = None,
+        dry_run: bool = False,
+        batch: bool = False,
+    ):
+        """
+        The driver.
+
+        :param cycle: The cycle.
+        :param config: Path to config file (read stdin if missing or None).
+        :param dry_run: Run in dry-run mode?
+        :param batch: Run component via the batch system?
+        """
+        super().__init__(config=config, dry_run=dry_run, batch=batch, cycle=cycle)
+        if self._dry_run:
+            dryrun()
+        self._cycle = cycle
+
+    # Workflow tasks
+
+    @task
+    def boundary_files(self):
+        """
+        Boundary files.
+        """
+        yield self._taskname("boundary files")
+        lbcs = self._driver_config["lateral_boundary_conditions"]
+        endhour = self._driver_config["length"]
+        interval = lbcs["interval_hours"]
+        symlinks = {}
+        for boundary_hour in range(0, endhour + 1, interval):
+            file_date = self._cycle + timedelta(hours=boundary_hour)
+            fn = f"lbc.{file_date.strftime('%Y-%m-%d_%H.%M.%S')}.nc"
+            linkname = self._rundir / fn
+            symlinks[linkname] = Path(lbcs["path"]) / fn
+        yield [symlink(target=t, linkname=l) for l, t in symlinks.items()]
+
+    @tasks
+    def files_copied(self):
+        """
+        Files copied for run.
+        """
+        yield self._taskname("files copied")
+        yield [
+            filecopy(src=Path(src), dst=self._rundir / dst)
+            for dst, src in self._driver_config.get("files_to_copy", {}).items()
+        ]
+
+    @tasks
+    def files_linked(self):
+        """
+        Files linked for run.
+        """
+        yield self._taskname("files linked")
+        yield [
+            symlink(target=Path(target), linkname=self._rundir / linkname)
+            for linkname, target in self._driver_config.get("files_to_link", {}).items()
+        ]
+
+    @task
+    def namelist_file(self):
+        """
+        The namelist file.
+        """
+        path = self._rundir / "namelist.atmosphere"
+        yield self._taskname(str(path))
+        yield asset(path, path.is_file)
+        yield None
+        duration = timedelta(hours=self._driver_config["length"])
+        str_duration = str(duration).replace(" days, ", "")
+        namelist = self._driver_config.get("namelist", {})
```

Comment:
I'm reviving this comment, which fell into "outdated" status and was hidden before we had a chance to discuss...

Thinking out loud: What is the case in which `namelist` is really optional here? Here are the cases I can think of:
1. User provides `namelist` with `updates_values` but no `base_file`. While possible, this may be unlikely because the `update_values` would have to completely specify the namelist content, which might mean a lot of YAML. But fine: We would still be able to add the cycle info and create the file.
2. User provides `namelist` with `base_file` but no `update_values`. Fine: We create an empty `update_values`, two lines down, populate it with the cycle info, and write the file.
3. User provides `namelist` with both `base_file` and `update_values`. Fine: We modify `update_values` with cycle info, and write the file.
4. User provides no `namelist`. That's permitted by the schema, but means that they must provide a complete, ready-to-go namelist file. If they do that, we'll never reach this code because `yield asset(path, path.is_file)`, above, will have located that file and this task will be deemed complete already.

Based on that case analysis, I wonder if we should use `self._driver_config["namelist"]` in a `try` / `except` block (or not provide the `{}` default value here and check for a `None` value) and raise a `UWConfigError` describing the situation. It seems to me like the `namelist` block is only optional in one case -- a case that should prevent execution of this code. I think we can detect if the user has done that incorrectly and help them here.

Please let me know if you see problems with that argument. But if it's right, we might need to open a ticket to make the same improvement in other drivers.

EDIT: What I'm proposing (here and similarly in the init driver) would be something like
```
try:
    namelist = self._driver_config("namelist")
except KeyError:
    raise UWConfigError("Provide either a 'namelist' YAML block or the %s file" % path)
```

Reply:
Just pushed this change in the two drivers. I still need to update the schema, though.

Reply:
Exactly. The schema needs to be permissive, but in the driver we know all the facts and can check things that we couldn't check at schema-validation time. (We don't explicitly check that the namelist file exists, but it's implied by the fact that the task method makes it past the `yield` statements and into the imperative code.)